### PR TITLE
Silent compiler warnings via -isystem includes

### PR DESCRIPTION
**Pull request by rhaschke**
rviz' [CI is frequently failing](https://github.com/ros-visualization/rviz/pull/1617) because of compiler warnings in include files external to the actual source tree. While Debian Buster uses an older gcc version and [thus isn't affected](https://build.ros.org/job/Npr_db__rviz__debian_buster_amd64/52/), warnings on the [ROS buildfarm for Ubuntu Focal](https://build.ros.org/job/Npr__rviz__ubuntu_focal_amd64/56) turn into failures in github's CI.

The obvious solution is to mark external includes using `-isystem` instead of `-I`. However, this causes issues if `/usr/include` is included like this as well - either directly or indirectly via relative paths or symlinks. Hence, the second change of this PR filters out these paths.

### CODE REVIEW

File: cmake/sip_helper.cmake [hunk 1]
```diff
@@ -92,6 +99,7 @@ function(build_sip_binding PROJECT_NAME SIP_FILE)
     add_custom_command(
         OUTPUT ${SIP_BUILD_DIR}/Makefile
         COMMAND ${PYTHON_EXECUTABLE} ${sip_SIP_CONFIGURE} ${SIP_BUILD_DIR} ${SIP_FILE} ${sip_LIBRARY_DIR} \"${INCLUDE_DIRS}\" \"${LIBRARIES}\" \"${LIBRARY_DIRS}\" \"${LDFLAGS_OTHER}\" \"${EXTRA_DEFINES}\"
+        COMMAND sed -i 's/ -I/ -isystem/g' ${SIP_BUILD_DIR}/Makefile
```
Comment (sloretz) [inline cmake/sip_helper.cmake:102]: I think there are windows users, and it seems like requiring `sed` would be a problem there. Can we do this in `sip_configure.py` [after generating the makefile](https://github.com/ros-visualization/python_qt_binding/blob/fd4d50b6e2b65e329039fba033cae406d8e3ad1a/cmake/sip_configure.py#L191)?

File: cmake/sip_helper.cmake [hunk 2]
```diff
@@ -75,7 +75,14 @@ function(build_sip_binding PROJECT_NAME SIP_FILE)
 
     set(SIP_BUILD_DIR ${sip_BINARY_DIR}/sip/${PROJECT_NAME})
 
-    set(INCLUDE_DIRS ${${PROJECT_NAME}_INCLUDE_DIRS} ${PYTHON_INCLUDE_DIRS})
+    set(INCLUDE_DIRS)
+    foreach(dir ${${PROJECT_NAME}_INCLUDE_DIRS} ${PYTHON_INCLUDE_DIRS})
+      get_filename_component(dir_real "${dir}" REALPATH BASE_DIR ${sip_SOURCE_DIR})
+      # filter out /usr/include, which must not be included via -isystem
+      if (NOT "${dir_real}" STREQUAL "/usr/include")
```
Comment (sloretz) [inline cmake/sip_helper.cmake:82]: I don't understand why filtering `/usr/include` is necessary, but I'm willing to try it in Noetic and see if it causes issues.

**Review by sloretz — CHANGES_REQUESTED** [review]
> The obvious solution is to mark external includes using -isystem instead of -I. However, this causes issues if /usr/include is included like this as well - either directly or indirectly via relative paths or symlinks. 

What happens if `/usr/include` is included as an `-isystem` include?

**Comment by rhaschke** [comment]
> What happens if `/usr/include` is included as an `-isystem` include?

Sorry for not providing this info in first place. See e.g. [here on stackoverflow](https://stackoverflow.com/questions/37218953/isystem-on-a-system-include-directory-causes-errors).

File: cmake/sip_helper.cmake (hunk 1 above)
Comment (rhaschke) [reply cmake/sip_helper.cmake:102]: Sure that will be possible. But, I will not have cycles to do that in any new future. I am using the present patch since several months already and just barely managed to file the PR now.

**Comment by wjwwood** [comment]
I'll just mention that we've had issues in the past with `-isystem` being used in combination with `-I` in the past. Basically if it was `-IA -IB -IC` and `foo.hpp` existed in all three, you'd take it from `-IC`, but if you changed it to `-IA -isystemB -IC` then it might change which version of `foo.hpp` was selected. I'm not sure that's exactly right (might have the scenario a bit wrong), and it only came up in overlaying scenarios, but we did have problems with it at some point and actually removed the use of `-isystem` in a lot of places.

This was years and years ago, so maybe it's worth doing now, but I just wanted to bring it up.

After looking, I guess @sloretz you're already aware of this, based on this: https://github.com/ros2/pybind11_vendor/issues/9

And it leads to somewhat fragile proposed workarounds like this: https://github.com/ros2/rosbag2/pull/623

Basically `-isystem` has been really annoying in the past.

It's more annoying to do so, but I think disabling warnings around include statements within rviz's code base is safer, but that's definitely a personal preference kind of choice.

**Comment by rhaschke** [comment]
@wjwwood, is the problem explained by this note in the [gcc docs](https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html)?

> If a standard system include directory, or a directory specified with `-isystem`, is also specified with `-I`, the `-I` option is _ignored_. The directory is still searched but as a system directory at its normal position in the system include chain.

This suggests that if you specify an include path both as `-I` and `-isystem`, it will be ignored for the high-priority `-I` search and only considered later when processing all `-isystem` includes. For me, that reads like _all_ external include paths should be marked consequently with `-isystem`.

**Comment by wjwwood** [comment]
It could be that using `-isystem` for "everything external" could work, but that's not the case right now. Using it sometimes is the issue I think though.

**Comment by sloretz** [comment]
> It could be that using -isystem for "everything external" could work, but that's not the case right now. Using it sometimes is the issue I think though.

IIUC in ROS 2 everything is `-isystem` because that's the default for imported CMake targets.

Since this PR is ROS 1, IIRC catkin uses CMake variables and does some workspace aware include directory ordering. The order in the gcc docs means that probably breaks when `-I` or `-isystem` are mixed.

> Sorry for not providing this info in first place. See e.g. here on stackoverflow.

Ah got it. It seems to be this issue: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70129 . CMake itself works around that as of https://gitlab.kitware.com/cmake/cmake/-/merge_requests/2716 . One comments says appending `/usr/include` to `CMAKE_CXX_IMPLICIT_INCLUDE_DIRECTORIES` will work around the issue in the meantime.

**Comment by rhaschke** [comment]
Note that the `sip` stuff doesn't use `cmake`. It directly creates a `make` file and this PR suggests replacing all `-I` with `-isystem`. Hence, there is no risk to mix them - everything is `-isystem` in the end (only for compiling the sip bindings). So, I think it would be safe to apply this PR. Of course, `sed` should be replaced by python...

**Comment by sloretz** [comment]
ROS 1 is end-of-life (EOL) as May 31st 2025, so I'm closing this.

[If you still rely on ROS 1, read this page to learn about your options.](https://bit.ly/NoeticEOL)
